FPU store/load for arm64 fastcontext

stelleg opened the pull request:
This addresses issue #123 by storing/loading FPU registers in the arm64 fastcontext.

janciesko commented:
CI testing is failing likely due to https://github.com/Qthreads/qthreads/issues/126.

olivier-snl commented:
It's great to have a potential fix to the bug.  We'll want to do broad testing.  In particular I'm interested to see if we need to do anything differently to handle ThunderX2 vs A64FX vs arm mac.  The reproducer derived from the failed chapel test happens to have a double time value.  I think a good unit test should be constructed to have some intentional FP work (including vector FP), then yield, then more FP.

cjhackillinois commented on include/fastcontext/arm-ucontext.h:
```diff
@@ -19,7 +19,7 @@ typedef struct uctxt uctxt_t;
     /* Saved main processor registers. */
 #ifdef NEEDARMA64CONTEXT
-	uint64_t regs[32]; /* callee saves x0-x30, SP */
+	uint64_t regs[40]; /* callee saves x0-x30, SP, FPU:d8-d16 */
```
Why 8 FPU registers? There are a total of 32 for Aarch64, no?

cjhackillinois commented on src/fastcontext/asm.S:
```diff
@@ -883,12 +889,18 @@ SET:
+        ldp     X29, X30, [X0,#232]
+
+        // FPU
+        ldp     D8,  D9,  [X0, #264]
```
Why start at offset of 264 instead of 256? 32 8-bytes registers is 256 bytes, so I would assume you could start saving the D registers at offset 256. Also, why these 8 registers?

cjhackillinois reviewed (commented):
I would like to see a reference why only 8 FPU registers are saved out of the 32? If a double is used, does gcc put the double into D8 as opposed to D0 or D1?

stelleg commented:
Some responses and updates to the helpful feedback from @cjhackillinois:

So the changes I had added above were borrowed from here: https://github.com/pmodels/argobots/blob/main/src/arch/fcontext/fcontext_arm64_aapcs_elf_gas.S. I thought maybe they knew something we didn't, but upon further inspection, I think the existing code that was ifdef'd out as NEON_REGS is doing the more complete FPU context swap. I've reverted my changes and replaced them with a simple removal of the NEON_REGS ifdef and a minor buffer change, because as I understand it those are non-optional in ARM64 and required for floating point, so I don't think it makes sense to have them be ifdef'd out.

janciesko commented:
@stelleg - Could you please rebase your PR onto the current main branch? That should make CI testing pass.

cjhackillinois commented on include/fastcontext/arm-ucontext.h:
```diff
@@ -20,9 +20,7 @@ struct mctxt {
-	#ifdef ARMA64_CONTEXT_SWITCH_NEON_REGS
-	uint128_t regs[32]; /* SIMD Neon Registers*/
-	#endif
+	uint64_t fpu_regs[64]; /* 32 128bit FPU/SIMD Neon Registers */
```
I think this could use the uint128_t type.

stelleg replied:
My understanding is that uint128_t is not required by the c specification, and often isn't defined. It wasn't for the compiler/math.h I was using when developing.

dmk42 replied:
> My understanding is that uint128_t is not required by the c specification, and often isn't defined. It wasn't for the compiler/stdint.h I was using when developing.

That is correct.  Only up to uint64_t is required by the standard (and only if the target supports it).  The Mac is one of the reasons for this PR, and it does not have uint128_t.

dmk42 replied:
FYI, C23 will require exact-width types for any width supported by the implementation, so in the future, if a 128-bit integer type is available, then uint128_t will be required.  C23 has not been published yet, though.

cjhackillinois replied:
I believe I tested with "qthread/qthread-int.h". That header file is generated by `./configure`.

ronawho commented:
Wohoo, very excited to see this! Both of the Chapel tests noted in https://github.com/Qthreads/qthreads/issues/113 that were failing do use floating points. I have confirmed those now run correctly on thunderx2.

If there's interest, I could run more comprehensive Chapel testing on thunderx2, a64fx, m1, and graviton.

Could also see if there's much difference in task creation/switching speed before and after this if that's a concern.

olivier-snl commented:
@ronawho on the mac side it seems that there may have been recent apple changes that will mean a little additional fix on that platform.  @janciesko reported that this may be Ventura-related.

We don't have access to graviton (that I know of).

ronawho commented:
> @ronawho on the mac side it seems that there may have been recent apple changes that will mean a little additional fix on that platform. @janciesko reported that this may be Ventura-related.

Just curious, Is that with r17 not being available or something else?

> We don't have access to graviton (that I know of).

I meant, I have access to those 4 platforms if you'd like me to run some testing on any of them.

janciesko commented:
On Ventura, it seems that autoconf is not correctly detecting AArch64 and thus sets `asssembly architecture` to  `UNSUPPORTED`.
